lib/livebook/hubs/enterprise_client.ex

```diff
@@ -110,31 +119,82 @@ defmodule Livebook.Hubs.EnterpriseClient do
   end
 
   def handle_info({:connect, :error, reason}, state) do
     Broadcasts.hub_connection_failed(reason)
     {:noreply, %{state | connected?: false, connection_error: reason}}
   end
 
-  def handle_info({:event, :secret_created, %{name: name, value: value}}, state) do
-    secret = %Secret{name: name, value: value, origin: {:hub, state.hub.id}}
+  def handle_info({:disconnect, :ok, :disconnected}, state) do
+    Broadcasts.hub_disconnected()
+    {:stop, :normal, state}
+  end
+
+  def handle_info({:event, topic, data}, state) do
+    Logger.debug("Received event #{topic} with data: #{inspect(data)}")
+
+    {:noreply, handle_event({topic, data}, state)}
+  end
+
+  def handle_event({:secret_created, secret_created}, state) do
+    secret = build_secret(state, secret_created)
     Broadcasts.secret_created(secret)
 
-    {:noreply, put_secret(state, secret)}
+    put_secret(state, secret)
   end
 
-  def handle_info({:event, :secret_updated, %{name: name, value: value}}, state) do
-    secret = %Secret{name: name, value: value, origin: {:hub, state.hub.id}}
+  def handle_event({:secret_updated, secret_updated}, state) do
+    secret = build_secret(state, secret_updated)
     Broadcasts.secret_updated(secret)
 
-    {:noreply, put_secret(state, secret)}
+    put_secret(state, secret)
   end
 
-  def handle_info({:disconnect, :ok, :disconnected}, state) do
-    Broadcasts.hub_disconnected()
-    {:stop, :normal, state}
+  def handle_event({:secret_deleted, secret_deleted}, state) do
+    secret = build_secret(state, secret_deleted)
+    Broadcasts.secret_deleted(secret)
+
+    remove_secret(state, secret)
+  end
+
+  def handle_event({:user_synchronized, user_synchronized}, %{secrets: []} = state) do
+    state = update_hub(state, user_synchronized.name)
+    secrets = for secret <- user_synchronized.secrets, do: build_secret(state, secret)
+
+    %{state | secrets: secrets}
+  end
+
+  def handle_event({:user_synchronized, user_synchronized}, state) do
+    state = update_hub(state, user_synchronized.name)
+    secrets = for secret <- user_synchronized.secrets, do: build_secret(state, secret)
+
+    created_secrets =
+      Enum.reject(secrets, fn secret ->
+        Enum.find(state.secrets, &(&1.name == secret.name and &1.value == secret.value))
+      end)
+
+    deleted_secrets =
+      Enum.reject(state.secrets, fn secret ->
+        Enum.find(secrets, &(&1.name == secret.name))
+      end)
+
+    updatd_secrets =
+      Enum.filter(secrets, fn secret ->
+        Enum.find(state.secrets, &(&1.name == secret.name and &1.value != secret.value))
+      end)
+
+    for secret <- deleted_secrets,
+        do: send(self(), {:event, :secret_deleted, secret})
+
+    for secret <- created_secrets,
+        do: send(self(), {:event, :secret_created, secret})
+
+    for secret <- updatd_secrets,
+        do: send(self(), {:event, :secret_updated, secret})
+
+    state
```

Comment:
Don't we need a catch-all `handle_event(_, _, _)` so we don't crash if we run against a server that sends more events that we know?

Reply:
Ideally we would log on the catch-all.